### ISSUE TITLE
[API Proposal]: Nullable<T> implement IEquatable<Nullable<T>>

### ISSUE THREAD

**Issue by ronimizy** [dotnet/runtime#87379]
### Background and motivation

As `Nullable<T>` does not have typed comparison method, only one that accepts `object?`, comparing with other `Nullable<T>` instance results in boxing of second comparison operand.

https://github.com/dotnet/runtime/blob/362678a6c7d45f9a49c4aaa946701f791c56b484/src/libraries/System.Private.CoreLib/src/System/Nullable.cs#L58-L63

`Nullable` static class already has an implementation, that avoids boxing.

https://github.com/dotnet/runtime/blob/362678a6c7d45f9a49c4aaa946701f791c56b484/src/libraries/System.Private.CoreLib/src/System/Nullable.cs#L90-L99

But using `Nullable.Equals(a, b)` is clunkier than `a.Equals(b)`, and second option is more commonly used (personally, never seen the first option in code bases), as most of developers probably does not event think that there is a boxing (I stumbled on it, just because I installed plugin on my IDE that indicated that there is boxing).

### API Proposal

```csharp
namespace System;

public partial struct Nullable<T> : IEquatable<Nullable<T>> where T : struct
{
    public bool Equals(Nullable<T> other);
}
```

### API Usage

```csharp
public void A(Guid? first, Guid? second)
{
    if (first.Equals(second))
    {
        Console.WriteLine("No boxing!");
    }
}
```

### Alternative Designs

Add `Equals(Nullable<T> other)` method without interface implementation.

### Risks

Possible stack overflow when comparing extremely large structs (highly unlikely)

**Comment by stephentoub** [comment]
https://github.com/dotnet/runtime/blob/362678a6c7d45f9a49c4aaa946701f791c56b484/src/libraries/System.Private.CoreLib/src/System/Nullable.cs#L12-L16

**Comment by ronimizy** [comment]
@stephentoub what about implementation described in alternative designs section?

**Comment by stephentoub** [comment]
Personally, I'd find it strange calling an instance Equals method on something that might be null. It overrides the base Equals(object) because it has to in order to have correct semantics, including supporting comparing with another null or T or boxed T.

**Comment by ronimizy** [comment]
@stephentoub In my experience, it is quite common to compare nullable values, ex: optional entity identifiers. I'm not suggesting to remove existing `Equals(object)` method, just adding a new one next to it, it will not break existing code, it will only reduce allocations on boxing for described usages.
If the case was so strange, why does method `Nullable.Equals` exist in the first place?

**Comment by stephentoub** [comment]
> If the case was so strange, why does method Nullable.Equals exist in the first place?

StringComparer.Ordinal.Equals(s1, s2) allows s1 to be null, while trying to call s1.Equals(s2) will throw an exception. I'm saying I'd similarly expect nullable1.Equals(nullable2) to throw an exception if nullable1 is null, but that's likely not what you want, nor is it consistent with object.Equals, in which case I'd personally rather it not exist.

**Comment by ronimizy** [comment]
> while trying to call s1.Equals(s2) will throw an exception

@stephentoub 

Yes, it will throw `NullReferenceException`, but `Nullable<T>` used with value types, thus we talk about value semantics, not reference semantics. 
It would be strange to expect `NullReferenceException` when doing something with value type instance.

Compliance with `object.Equals` semantics is different question, as `Nullable<T>.Equals` already does not, there will be no `NullReferenceException` if method called on object where `HasValue` is false.

My issue is purely about performance and allocations, proposed change (implementation in alternative designs section) does not affect existing semantics in any way.

**Comment by stephentoub** [comment]
> but Nullable<T> used with value types, thus we talk about value semantics, not reference semantics

The purpose of nullable is to give value types reference-like behavior regarding nulls. It's why Value throws an exception if the value is null.

> Compliance with object.Equals semantics is different question, as Nullable<T>.Equals already does not, there will be no NullReferenceException if method called on object where HasValue is false.

I called out the object.Equals override because without the override, calling Equals on a nullable would be semantically broken. I think it's a mistake that it doesn't throw for this == null. 

My original comment was highlighting that we can't do the primary proposal in this issue. You then asked for my opinion on the alternative, and 
I shared my personal opinion.

**Comment by ronimizy** [comment]
@stephentoub so what should I do in this case? Close this proposal and open the new one described in alternative implementation section?

**Comment by teo-tsirpanis** [comment]
@ronimizy you could modify this issue instead of opening another one.

**Comment by ronimizy** [comment]
@stephentoub modified the issue, removed mentions of interface implementation

**Comment by ronimizy** [comment]
@stephentoub 👀

**Comment by AlexRadch** [comment]
How about supporting 4 different `Nullable<T>` within the runtime and compiler?
1. If T has no interfaces, use the regular `Nullable<T>`.
2. If T has IEquatable<T> interface, than use `EquatableNullable<T>` that support `IEquatable<EquatableNullable<T>>` interface.
3. If T has IComparable<T> interface, than use `ComparableNullable<T>` that support `IComparable<ComparableNullable<T>>` interface.
4. If T has IEquatable<T> and IComparable<T> interfaces, than use `EquatableAndComparableNullable<T>` that support `IEquatable<EquatableAndComparableNullable<T>>` and `IComparable<EquatableAndComparableNullable<T>>` interfaces.

**Comment by DaZombieKiller** [comment]
> but taking into account the fact that Span now does not support nullable types because of this, it is better this way than the way it is now.

`Span<T>` does support nullable types. Can you elaborate on what you mean by this?

**Comment by AlexRadch** [comment]
> `Span<T>` does support nullable types. Can you elaborate on what you mean by this?

Try to perform at least one useful method on a span with a nullable structure (numbers, structures, records, etc.), find an element in a list, sort a list, perform a binary search, etc. These operations are not available, although arrays support them, but spans - no.

**Comment by DaZombieKiller** [comment]
Those are all `MemoryExtensions` methods, `Span<T>` itself supports nullable just fine. You most likely want https://github.com/dotnet/runtime/issues/75639 which would unblock operations like `IndexOf` for `Nullable<T>` and other types that don't directly implement `IEquatable<T>`.

**Comment by tannergooding** [comment]
Nullable is special and as Stephen pointed out we can't implement an interface on it. We could expose the method without the interface, but that's likely not worthwhile over just using the existing language features or adding an extension method for it if you need support in your own library.
